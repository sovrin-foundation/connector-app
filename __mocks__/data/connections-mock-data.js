// @flow

export const botLogoUrl = 'http://cr0ybot.github.io/ingress-logos/ingress.png'
export const connectionThemes = {
  [botLogoUrl]: {
    primary: 'rgba(100,200,100,1)',
    secondary: 'rgba(100,200,100,0.6)',
  },
}
