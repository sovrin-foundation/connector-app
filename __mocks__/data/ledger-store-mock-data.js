// @flow
import { STORE_STATUS } from '../../app/common/type-common'

export const transferFees = {
  transfer: '0.00001',
}

export const ledgerStoreWithTransferFees = {
  fees: {
    data: transferFees,
    status: STORE_STATUS.SUCCESS,
    error: null,
  },
}
