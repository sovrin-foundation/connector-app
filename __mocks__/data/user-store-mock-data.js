// @flow

export const userOneTimeInfo = {
  oneTimeAgencyDid: 'oneTimeAgencyDid',
  oneTimeAgencyVerificationKey: 'oneTimeAgencyVerKey',
  myOneTimeDid: 'myOneTimeDid',
  myOneTimeVerificationKey: 'myOneTimeVerificationKey',
  myOneTimeAgentDid: 'myOneTimeAgentDid',
  myOneTimeAgentVerificationKey: 'myOneTimeAgentVerificationKey',
}
