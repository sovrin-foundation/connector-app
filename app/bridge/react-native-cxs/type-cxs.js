// @flow

import type { SMSPendingInvitationPayload } from '../../sms-pending-invitation/type-sms-pending-invitation'
import type { UserOneTimeInfo } from '../../store/user/type-user-store'
import type { AgencyPoolConfig } from '../../store/type-config-store'
import type { GenericObject } from '../../common/type-common'
import type {
  ClaimOfferPushPayload,
  ClaimPushPayload,
} from '../../push-notification/type-push-notification'
import type { ProofRequestPushPayload } from '../../proof-request/type-proof-request'

export type Metadata = {
  [string]: any,
}

export type IndyClaimOffer = {
  issuerDid: string,
  schemaSequenceNumber: number,
}

export type IndyConnectionHistory = {
  issuerDid: string,
  schemaSequenceNumber: number,
}

export type IndyClaimRequest = {
  blinded_ms: {
    prover_did: string,
    u: string,
    ur?: string,
  },
  issuer_did: string,
  schema_seq_no: number,
}

export type ConnectAgencyResponse = {
  '@type': {
    name: string,
    ver: string,
  },
  withPairwiseDID: string,
  withPairwiseDIDVerKey: string,
}

export type RegisterAgencyResponse = {
  '@type': { name: string, ver: string },
}

export type CreateOneTimeAgentResponse = {
  '@type': { name: string, ver: string },
  withPairwiseDID: string,
  withPairwiseDIDVerKey: string,
}

export type CreatePairwiseAgentResponse = {
  '@type': { name: string, ver: string },
  withPairwiseDID: string,
  withPairwiseDIDVerKey: string,
}

export type AcceptInvitationResponse = {
  '@type': { name: string, ver: string },
  uid: string,
}

// TODO: Remove types that will not be used once vcx is integrated
export type VcxProvision = {
  agency_url: string,
  agency_did: string,
  agency_verkey: string,
}

export type VcxProvisionResult = {
  wallet_name: string,
  wallet_key: string,
  agency_endpoint: string,
  agency_did: string,
  agency_verkey: string,
  // myOneTimeDid
  sdk_to_remote_did: string,
  // myOneTimeVerificationKey
  sdk_to_remote_verkey: string,
  // oneTimeAgencyDid
  institution_did: string,
  // oneTimeAgencyVerificationKey
  institution_verkey: string,
  // myOneTimeAgentDid
  remote_to_sdk_did: string,
  // myOneTimeAgentVerificationKey
  remote_to_sdk_verkey: string,
}

export type CxsInitConfig = UserOneTimeInfo & AgencyPoolConfig

export type InitWithGenesisPathConfig = CxsInitConfig & { genesis_path: string }

export type VcxInitConfig = {
  agency_endpoint: string,
  agency_did: string,
  agency_verkey: string,
  genesis_path: string,
  wallet_key: string,
  config: string,
  pool_name: string,
  wallet_name: string,
  remote_to_sdk_did: string,
  remote_to_sdk_verkey: string,
  sdk_to_remote_did: string,
  sdk_to_remote_verkey: string,
  institution_did: string,
  institution_verkey: string,
}

export type VcxPushTokenConfig = {
  id: string,
  value: string,
}

export type CxsPushTokenConfig = {
  uniqueId: string,
  pushToken: string,
}

export type VcxCreateConnection = {
  source_id: string,
  invite_details: SMSPendingInvitationPayload,
}

export type VcxConnectionCreateResult = number

export type VcxConnectionConnectResult = {
  source_id: string,
  pw_did: string,
  pw_verkey: string,
  uuid: string,
  endpoint: string,
  agent_did: string,
  agent_vk: string,
  their_pw_did: string,
  their_pw_verkey: string,
}

export type VcxCredentialOfferResult = {
  credential_handle: number,
  credential_offer: string,
}

export type VcxCredentialOffer = {
  msg_type: string,
  version: string,
  to_did: string,
  from_did: string,
  libindy_offer: string,
  cred_def_id: string,
  credential_attrs: { [string]: Array<string> },
  schema_seq_no: number,
  claim_name: string,
  claim_id: string,
  payment_address?: ?string,
  price?: ?string,
}

export type CxsCredentialOfferResult = {
  claimHandle: number,
  claimOffer: ClaimOfferPushPayload,
}

export type VcxSendCredentialRequest = {}

export type VcxClaimInfo = {
  credential_id?: string,
  credential?: ClaimPushPayload,
  credential_offer?: string,
}

export type VcxProofRequest = ProofRequestPushPayload

export type WalletPoolName = {
  walletName: string,
  poolName: string,
}

export type UTXO = {
  amount: number,
  extra: string,
  paymentAddress: string,
  source: string,
}

export type PaymentAddress = {
  address: string,
  balance: number,
  utxo: UTXO[],
}

export type WalletTokenInfo = {
  balance_str: string,
  addresses: PaymentAddress[],
}

export const smallDeviceMemory = 1073741824
