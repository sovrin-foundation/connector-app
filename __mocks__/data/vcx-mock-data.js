// @flow

import {
  agencyUrl,
  agencyDID,
  agencyVerificationKey,
} from './config-store-mock-data'
import { userOneTimeInfo } from './user-store-mock-data'
import { WALLET_FILE_NAME } from '../../app/backup/type-backup'

// I believe now is the time when we should start splitting static-data.js
// we are starting to see that we shy away from utilizing existing data,
// which was the intention of creating one file, so that same object
// would be used all throughout app and we would know breakage of contract
// among screens/components as soon as we change some data at one place.
// `flow` solves most of those issues, but still there are some issues
// that flow can't solve, and for those we need test to identify
// and for tests we need some mocked data which will be used throughout tests

export const vcxProvisionResult = {
  wallet_name: WALLET_FILE_NAME,
  wallet_key: 'walletKey',
  agency_endpoint: agencyUrl,
  agency_did: agencyDID,
  agency_verkey: agencyVerificationKey,
  sdk_to_remote_did: userOneTimeInfo.myOneTimeDid,
  sdk_to_remote_verkey: userOneTimeInfo.myOneTimeVerificationKey,
  institution_did: userOneTimeInfo.oneTimeAgencyDid,
  institution_verkey: userOneTimeInfo.oneTimeAgencyVerificationKey,
  remote_to_sdk_did: userOneTimeInfo.myOneTimeAgentDid,
  remote_to_sdk_verkey: userOneTimeInfo.myOneTimeAgentVerificationKey,
}

export const vcxConnectionSerialized = JSON.stringify({})

export const serializedClaimOffer = '{}'

export const serializedClaimOffers = JSON.stringify({
  pairwiseIdentifier1: {
    uid: serializedClaimOffer,
  },
})

export const vcxClaimOffer = {
  msg_type: 'CLAIM_OFFER',
  version: '1.0.0',
  to_did: '8XFh8yBzrpJQmNyZzgoTqB',
  from_did: 'ha66899sadfjZJGINKN0770',
  libindy_offer: '',
  cred_def_id: 'cred_def_id',
  credential_attrs: {
    ['Address 1']: ['Address Address Address'],
    ['Address 2']: ['Address 2 Address 2 Address 2'],
  },
  claim_name: 'Home Address',
  schema_seq_no: 36,
  claim_id: 'jhkad:97:kkda:jhh',
}
