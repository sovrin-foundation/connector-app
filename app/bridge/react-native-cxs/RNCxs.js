// @flow
import { NativeModules, Platform } from 'react-native'
import memoize from 'lodash.memoize'
import type {
  Metadata,
  IndyClaimOffer,
  ConnectAgencyResponse,
  RegisterAgencyResponse,
  CreateOneTimeAgentResponse,
  CreatePairwiseAgentResponse,
  AcceptInvitationResponse,
  CxsCredentialOfferResult,
  InitWithGenesisPathConfig,
  VcxProvisionResult,
  VcxProvision,
  CxsInitConfig,
  VcxInitConfig,
  CxsPushTokenConfig,
  VcxCredentialOfferResult,
  VcxCredentialOffer,
  VcxClaimInfo,
  VcxConnectionConnectResult,
  VcxProofRequest,
  WalletTokenInfo,
  PaymentAddress,
} from './type-cxs'
import type { InvitationPayload } from '../../invitation/type-invitation'
import {
  convertAgencyConfigToVcxProvision,
  convertVcxProvisionResultToUserOneTimeInfo,
  convertCxsInitToVcxInit,
  convertCxsPushConfigToVcxPushTokenConfig,
  convertInvitationToVcxConnectionCreate,
  convertVcxConnectionToCxsConnection,
  convertVcxCredentialOfferToCxsClaimOffer,
  paymentHandle,
  getWalletKey,
  convertSovrinTokensToSovrinAtoms,
  convertSovrinAtomsToSovrinTokens,
  convertVcxLedgerFeesToLedgerFees,
} from './vcx-transformers'
import type { UserOneTimeInfo } from '../../store/user/type-user-store'
import type { AgencyPoolConfig } from '../../store/type-config-store'
import type { MyPairwiseInfo } from '../../store/type-connection-store'
import type {
  ClaimOfferPushPayload,
  ClaimPushPayload,
} from '../../push-notification/type-push-notification'
import type {
  WalletHistoryEvent,
  WalletPayload,
} from '../../wallet/type-wallet'
import type { GenericStringObject } from '../../common/type-common'
import type { Passphrase } from '../../backup/type-backup'
import type { GetClaimVcxResult } from '../../claim/type-claim'
import uniqueId from 'react-native-unique-id'
import { smallDeviceMemory } from './type-cxs'
import { secureSet } from '../../services/storage'
import { __uniqueId } from '../../store/type-config-store'
import type { LedgerFeesData } from '../../store/ledger/type-ledger-store'

const { RNIndy } = NativeModules

export async function acceptInvitationVcx(
  connectionHandle: number
): Promise<MyPairwiseInfo> {
  // hard coding connection options to QR type for now, because vcx needs connection options
  // API for vcx assumes that it is running on enterprise side and not from consumer side
  // hence it tries to create connection with connection type.
  // However, our need is not to create a connection but to create a connection instance
  // with existing invitation. So, for now for any invitation type QR or SMS
  // we are hard coding connection option to QR
  const connectionOptions = { connection_type: 'QR', phone: '' }
  const result: string = await RNIndy.vcxAcceptInvitation(
    connectionHandle,
    JSON.stringify(connectionOptions)
  )
  // TODO:KS Remove below API call once sdk team returns pairwise info in above api
  // above call does not return pairwise did information, but we need pairwise info
  // to store that information and have those details available while making a connection
  // we have to make an extra call to get pairwise info
  const serializedConnection: string = await serializeConnection(
    connectionHandle
  )

  const vcxConnection: { data: VcxConnectionConnectResult } = JSON.parse(
    serializedConnection
  )

  return convertVcxConnectionToCxsConnection(vcxConnection.data)
}

export async function updatePushTokenVcx(pushTokenConfig: CxsPushTokenConfig) {
  return await RNIndy.vcxUpdatePushToken(
    JSON.stringify(convertCxsPushConfigToVcxPushTokenConfig(pushTokenConfig))
  )
}

export async function deleteConnection(connectionHandle: number) {
  return await RNIndy.deleteConnection(connectionHandle)
}

export async function resetVcx(removeData: boolean): Promise<boolean> {
  // we would remove above reset method and rename this method to reset
  // once we have integration available with vcx
  const result: boolean = await RNIndy.reset(true)

  return result
}

export async function vcxShutdown(deletePool: boolean): Promise<boolean> {
  const result: Number = await RNIndy.shutdownVcx(deletePool)

  return true
}

export async function getColor(imagePath: string): Promise<Array<string>> {
  return RNIndy.getColor(imagePath)
}

export async function sendTokenAmount(
  tokenAmount: string,
  recipientWalletAddress: string
): Promise<boolean> {
  const sovrinAtoms = convertSovrinTokensToSovrinAtoms(tokenAmount)
  return RNIndy.sendTokens(paymentHandle, sovrinAtoms, recipientWalletAddress)
}

export async function createOneTimeInfo(
  agencyConfig: AgencyPoolConfig
): Promise<UserOneTimeInfo> {
  const walletPoolName = await getWalletPoolName()
  const vcxProvisionConfig: VcxProvision = await convertAgencyConfigToVcxProvision(
    agencyConfig,
    walletPoolName
  )
  const provisionVcxResult: string = await RNIndy.createOneTimeInfo(
    JSON.stringify(vcxProvisionConfig)
  )
  const provisionResult: VcxProvisionResult = JSON.parse(provisionVcxResult)
  return convertVcxProvisionResultToUserOneTimeInfo(provisionResult)
}

export async function init(
  config: CxsInitConfig,
  fileName: string
): Promise<boolean> {
  const genesis_path: string = await RNIndy.getGenesisPathWithConfig(
    config.poolConfig,
    fileName
  )

  const initConfig = {
    ...config,
    genesis_path,
  }
  const walletPoolName = await getWalletPoolName()
  const vcxInitConfig: VcxInitConfig = await convertCxsInitToVcxInit(
    initConfig,
    walletPoolName
  )
  const initResult: boolean = await RNIndy.init(JSON.stringify(vcxInitConfig))

  return initResult
}

// TODO:KS Need to rename this to something like walletInit
export async function simpleInit(): Promise<boolean> {
  const walletPoolName = await getWalletPoolName()
  const wallet_key = await getWalletKey()
  const initConfig = {
    wallet_name: walletPoolName.walletName,
    wallet_key,
  }
  const initResult: boolean = await RNIndy.init(JSON.stringify(initConfig))
  return initResult
}

export const getWalletPoolName = memoize(async function() {
  const appUniqueId = await uniqueId()
  await secureSet(__uniqueId, appUniqueId)
  const walletName = `${appUniqueId}-cm-wallet`
  // Not sure why, but VCX is applying validation check on pool name
  // they don't like alphanumeric or _, so we have to remove "-"
  // from our guid that we generated
  const strippedAppUniqueId = appUniqueId.replace(/\-/g, '')
  const poolName = `${strippedAppUniqueId}cmpool`
  return {
    walletName,
    poolName,
  }
})

export async function createConnectionWithInvite(
  invitation: InvitationPayload
): Promise<number> {
  const { invite_details } = convertInvitationToVcxConnectionCreate(invitation)
  const connectionHandle: number = await RNIndy.createConnectionWithInvite(
    invitation.requestId,
    JSON.stringify(invite_details)
  )

  return connectionHandle
}

export async function serializeConnection(
  connectionHandle: number
): Promise<string> {
  const serializedConnection: string = await RNIndy.getSerializedConnection(
    connectionHandle
  )

  return serializedConnection
}

// cache Promise of serializedString so that we don't call Bridge method again
export const getHandleBySerializedConnection = memoize(async function(
  serializedConnection: string
): Promise<number> {
  const connectionHandle: number = await RNIndy.deserializeConnection(
    serializedConnection
  )

  return connectionHandle
})

export async function downloadClaimOffer(
  connectionHandle: number,
  sourceId: string,
  messageId: string
): Promise<CxsCredentialOfferResult> {
  const {
    credential_offer,
    credential_handle,
  }: VcxCredentialOfferResult = await RNIndy.credentialCreateWithMsgId(
    sourceId,
    connectionHandle,
    messageId
  )

  const {
    credential_offer: vcxCredentialOffer,
    price,
  }: { credential_offer: VcxCredentialOffer, price: number } = JSON.parse(
    credential_offer
  )

  vcxCredentialOffer.price = price
    ? convertSovrinAtomsToSovrinTokens(price)
    : null

  return {
    claimHandle: credential_handle,
    claimOffer: convertVcxCredentialOfferToCxsClaimOffer(vcxCredentialOffer),
  }
}

export async function serializeClaimOffer(
  claimHandle: number
): Promise<string> {
  const serializedClaimOffer: string = await RNIndy.serializeClaimOffer(
    claimHandle
  )

  return serializedClaimOffer
}

export const getClaimHandleBySerializedClaimOffer = memoize(async function(
  serializedClaimOffer: string
): Promise<number> {
  const claimHandle: number = await RNIndy.deserializeClaimOffer(
    serializedClaimOffer
  )

  return claimHandle
})

export async function sendClaimRequest(
  claimHandle: number,
  connectionHandle: number,
  paymentHandle: number
): Promise<void> {
  return await RNIndy.sendClaimRequest(
    claimHandle,
    connectionHandle,
    paymentHandle
  )
}

export async function downloadProofRequest(
  sourceId: string,
  connectionHandle: number,
  messageId: string
): Promise<VcxProofRequest> {
  const {
    proofHandle,
    proofRequest,
  }: {
    proofHandle: number,
    proofRequest: string,
  } = await RNIndy.proofCreateWithMsgId(sourceId, connectionHandle, messageId)

  const vcxProofRequest: VcxProofRequest = JSON.parse(proofRequest)

  return {
    ...vcxProofRequest,
    proofHandle,
  }
}

export async function getWalletBalance(): Promise<string> {
  const { balance_str: balance }: WalletTokenInfo = await getWalletTokenInfo()

  return convertSovrinAtomsToSovrinTokens(balance)
}

export async function getWalletAddresses(): Promise<string[]> {
  const { addresses }: WalletTokenInfo = await getWalletTokenInfo()

  // TODO:KS For now we don't need to store any other data on our react-native
  // when we need to store all data, then we would return only addresses
  // for now we are just returning addresses and ignoring anything else
  return addresses.map((address: PaymentAddress) => address.address)
}

export async function getWalletHistory(): Promise<WalletHistoryEvent[]> {
  const walletHistoryData = [
    {
      id: 'asd',
      senderAddress:
        'sov:ksudgyi8f98gsih7655hgifuyg79s89s98ydf98fg7gks8fjhkss8f030',
      action: 'Withdraw',
      tokenAmount: '5656',
      timeStamp: 'Tue, 04 Aug 2015 12:38:41 GMT',
    },
    {
      id: 'kld',
      senderName: 'Sovrin Foundation',
      senderAddress:
        'sov:ksudgyi8f98gsih7655hgifuyg79s89s98ydf98fg7gks8fjhkss8f030',
      action: 'Purchase',
      tokenAmount: '10000',
      timeStamp: 'Tue, 04 Aug 2015 14:38:41 GMT',
    },
  ]

  return new Promise.resolve(walletHistoryData)
}

export async function encryptWallet({
  encryptedFileLocation,
  recoveryPassphrase,
}: {
  encryptedFileLocation: string,
  recoveryPassphrase: Passphrase,
}): Promise<number> {
  const exportHandle: number = await RNIndy.exportWallet(
    encryptedFileLocation,
    recoveryPassphrase.hash
  )

  return exportHandle
}

export async function decryptWalletFile(
  walletPath: string,
  decryptionKey: string
): Promise<number> {
  const { walletName } = await getWalletPoolName()
  const wallet_key = await getWalletKey()

  const config = JSON.stringify({
    wallet_name: walletName,
    wallet_key,
    exported_wallet_path: walletPath,
    backup_key: decryptionKey,
  })
  const importHandle: number = await RNIndy.decryptWalletFile(config)

  return importHandle
}

export async function copyToPath(
  uri: string,
  destPath: string
): Promise<number> {
  return await RNIndy.copyToPath(uri, destPath)
}

export async function updateClaimOfferState(claimHandle: number) {
  const updatedState: number = await RNIndy.updateClaimOfferState(claimHandle)

  return updatedState
}

export async function getClaimOfferState(claimHandle: number): Promise<number> {
  const state: number = await RNIndy.getClaimOfferState(claimHandle)

  return state
}

export async function getClaimVcx(
  claimHandle: number
): Promise<GetClaimVcxResult> {
  const vcxClaimResult: string = await RNIndy.getClaimVcx(claimHandle)
  const vcxClaim: VcxClaimInfo = JSON.parse(vcxClaimResult)
  const { credential, credential_id } = vcxClaim

  if (!credential || !credential_id) {
    throw new Error('credential not found in vcx')
  }

  return {
    claimUuid: credential_id,
    claim: credential,
  }
}

export async function exportWallet(wallet: WalletPayload): Promise<number> {
  const exportHandle: number = await RNIndy.exportWallet(
    wallet.walletPath,
    wallet.encryptionKey
  )

  return exportHandle
}

export async function setWalletItem(
  key: string,
  value: string
): Promise<number> {
  return await RNIndy.setWalletItem(key, value)
}

export async function getWalletItem(key: string): Promise<string> {
  const response: string = await RNIndy.getWalletItem(key)
  if (response) {
    const itemValue = JSON.parse(response)
    const { value } = itemValue

    if (!value) {
      throw new Error('cannot get value')
    }

    return value
  } else {
    return response
  }
}

export async function deleteWalletItem(key: string): Promise<number> {
  return await RNIndy.deleteWalletItem(key)
}

export async function updateWalletItem(
  key: string,
  data: string
): Promise<number> {
  return await RNIndy.updateWalletItem(key, data)
}

export function exitAppAndroid() {
  if (Platform.OS === 'android') {
    RNIndy.exitAppAndroid()
  }
}

export async function getMatchingCredentials(
  proofHandle: number
): Promise<string> {
  // TODO:KS Add a transformer here, for now we are just passing through
  return await RNIndy.proofRetrieveCredentials(proofHandle)
}

export async function generateProof(
  proofHandle: number,
  selectedCredentials: string,
  selfAttestedAttributes: string
): Promise<void> {
  return await RNIndy.proofGenerate(
    proofHandle,
    selectedCredentials,
    selfAttestedAttributes
  )
}

export async function sendProof(
  proofHandle: number,
  connectionHandle: number
): Promise<void> {
  return await RNIndy.proofSend(proofHandle, connectionHandle)
}

export async function proofCreateWithRequest(
  sourceId: string,
  proofRequest: string
): Promise<number> {
  return await RNIndy.proofCreateWithRequest(sourceId, proofRequest)
}

export async function proofSerialize(proofHandle: number): Promise<string> {
  return await RNIndy.proofSerialize(proofHandle)
}

export async function proofDeserialize(
  serializedProofRequest: string
): Promise<number> {
  return await RNIndy.proofDeserialize(serializedProofRequest)
}

export async function downloadMessages(
  messageStatus: string,
  uid_s: ?string,
  pwdids: string
): Promise<string> {
  return await RNIndy.downloadMessages(messageStatus, uid_s, pwdids)
}

export async function updateMessages(
  messageStatus: string,
  pwdidsJson: string
): Promise<number> {
  return await RNIndy.updateMessages(messageStatus, pwdidsJson)
}

export async function getWalletTokenInfo(): Promise<WalletTokenInfo> {
  const paymentHandle = 0
  const tokenInfo: string = await RNIndy.getTokenInfo(paymentHandle)
  return JSON.parse(tokenInfo)
}

export async function createPaymentAddress(seed: ?string) {
  return RNIndy.createPaymentAddress(seed)
}

export async function getLedgerFees(): Promise<LedgerFeesData> {
  const fees: string = await RNIndy.getLedgerFees()

  return convertVcxLedgerFeesToLedgerFees(fees)
}

export const checkIfAnimationToUse = memoize(function() {
  return Platform.OS === 'android'
    ? RNIndy.totalMemory / smallDeviceMemory < 1 ? true : false
    : false
})

export async function getBiometricError(): Promise<string> {
  if (Platform.OS === 'ios') {
    return await RNIndy.getBiometricError()
  }
  return ''
}
