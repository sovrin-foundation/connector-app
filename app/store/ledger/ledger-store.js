// @flow

import { STORE_STATUS } from '../../common/type-common'
import {
  GET_LEDGER_FEES,
  getLedgerFeesFail,
  ERROR_GET_LEDGER_FEES,
  getLedgerFeesSuccess,
  GET_LEDGER_FEES_SUCCESS,
  GET_LEDGER_FEES_FAIL,
  RESET_LEDGER_FEES,
} from './type-ledger-store'
import type {
  LedgerAction,
  LedgerStore,
  LedgerFeesData,
} from './type-ledger-store'
import { put, take, all, call, select, takeLatest } from 'redux-saga/effects'
import { getLedgerFees } from '../../bridge/react-native-cxs/RNCxs'
import { ensureVcxInitSuccess } from '../config-store'
import { captureError } from '../../services/error/error-handler'

const initialState = {
  fees: {
    data: {
      transfer: '0',
    },
    status: STORE_STATUS.IDLE,
    error: null,
  },
}

export function* getLedgerFeesSaga(): Generator<*, *, *> {
  try {
    const vcxResult = yield* ensureVcxInitSuccess()
    if (vcxResult && vcxResult.fail) {
      throw new Error(JSON.stringify(vcxResult.fail.message))
    }
    const fees: LedgerFeesData = yield call(getLedgerFees)
    yield put(getLedgerFeesSuccess(fees))
  } catch (e) {
    captureError(e)
    yield put(getLedgerFeesFail(ERROR_GET_LEDGER_FEES(e.message)))
  }
}

export function* watchGetLedgerFees(): any {
  yield takeLatest(GET_LEDGER_FEES, getLedgerFeesSaga)
}

export function* watchLedgerStore(): any {
  yield all([watchGetLedgerFees()])
}

export function ledgerStoreReducer(
  state: LedgerStore = initialState,
  action: LedgerAction
): LedgerStore {
  switch (action.type) {
    case GET_LEDGER_FEES:
      return {
        ...state,
        fees: {
          ...state.fees,
          status: STORE_STATUS.IN_PROGRESS,
          error: null,
        },
      }

    case GET_LEDGER_FEES_SUCCESS:
      return {
        ...state,
        fees: {
          ...state.fees,
          status: STORE_STATUS.SUCCESS,
          error: null,
          data: {
            ...action.fees,
          },
        },
      }

    case GET_LEDGER_FEES_FAIL:
      return {
        ...state,
        fees: {
          ...state.fees,
          status: STORE_STATUS.ERROR,
          error: action.error,
        },
      }

    case RESET_LEDGER_FEES:
      return initialState

    default:
      return state
  }
}
