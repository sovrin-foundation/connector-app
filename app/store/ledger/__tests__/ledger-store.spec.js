// @flow

import { ledgerStoreReducer, getLedgerFeesSaga } from '../ledger-store'
import {
  getLedgerFees,
  getLedgerFeesSuccess,
  ERROR_GET_LEDGER_FEES,
  getLedgerFeesFail,
  resetLedgerFees,
} from '../type-ledger-store'
import { transferFees } from '../../../../__mocks__/data/ledger-store-mock-data'
import { configStoreHydratedInstalledVcxInitSuccess } from '../../../../__mocks__/data/config-store-mock-data'
import { expectSaga } from 'redux-saga-test-plan'
import * as matchers from 'redux-saga-test-plan/matchers'
import { throwError } from 'redux-saga-test-plan/providers'
import { getLedgerFees as getLedgerFeesApi } from '../../../bridge/react-native-cxs/RNCxs'

describe('ledgerStore', () => {
  it('action: GET_LEDGER_FEES', () => {
    expect(ledgerStoreReducer(undefined, getLedgerFees())).toMatchSnapshot()
  })

  it('action: GET_LEDGER_FEES_SUCCESS', () => {
    const stateAfterLedgerFeesStart = ledgerStoreReducer(
      undefined,
      getLedgerFees()
    )
    expect(
      ledgerStoreReducer(
        stateAfterLedgerFeesStart,
        getLedgerFeesSuccess(transferFees)
      )
    ).toMatchSnapshot()
  })

  it('action: GET_LEDGER_FEES_FAIL', () => {
    const stateAfterLedgerFeesStart = ledgerStoreReducer(
      undefined,
      getLedgerFees()
    )
    const error = ERROR_GET_LEDGER_FEES('Test error')
    expect(
      ledgerStoreReducer(stateAfterLedgerFeesStart, getLedgerFeesFail(error))
    ).toMatchSnapshot()
  })

  it('action: RESET_LEDGER_FEES', () => {
    const stateAfterLedgerFeesStart = ledgerStoreReducer(
      undefined,
      getLedgerFees()
    )
    const stateAfterSuccess = ledgerStoreReducer(
      stateAfterLedgerFeesStart,
      getLedgerFeesSuccess(transferFees)
    )
    expect(
      ledgerStoreReducer(stateAfterSuccess, resetLedgerFees())
    ).toMatchSnapshot()
  })

  it('saga: getLedgerFeesSaga => success', () => {
    const stateWithVcxSuccess = {
      config: configStoreHydratedInstalledVcxInitSuccess,
    }

    return expectSaga(getLedgerFeesSaga, getLedgerFees())
      .withState(stateWithVcxSuccess)
      .provide([[matchers.call.fn(getLedgerFeesApi), transferFees]])
      .put(getLedgerFeesSuccess(transferFees))
      .run()
  })

  it('saga: getLedgerFeesSaga => fail', () => {
    const stateWithVcxSuccess = {
      config: configStoreHydratedInstalledVcxInitSuccess,
    }
    const errorMessage = 'test error'
    const error = new Error(errorMessage)

    return expectSaga(getLedgerFeesSaga, getLedgerFees())
      .withState(stateWithVcxSuccess)
      .provide([[matchers.call.fn(getLedgerFeesApi), throwError(error)]])
      .put(getLedgerFeesFail(ERROR_GET_LEDGER_FEES(errorMessage)))
      .run()
  })
})
