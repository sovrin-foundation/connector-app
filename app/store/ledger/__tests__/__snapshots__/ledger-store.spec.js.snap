// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ledgerStore action: GET_LEDGER_FEES 1`] = `
Object {
  "fees": Object {
    "data": Object {
      "transfer": "0",
    },
    "error": null,
    "status": "IN_PROGRESS",
  },
}
`;

exports[`ledgerStore action: GET_LEDGER_FEES_FAIL 1`] = `
Object {
  "fees": Object {
    "data": Object {
      "transfer": "0",
    },
    "error": Object {
      "code": "LE-001",
      "message": "Error occurred while fetching ledger fees: Test error",
    },
    "status": "ERROR",
  },
}
`;

exports[`ledgerStore action: GET_LEDGER_FEES_SUCCESS 1`] = `
Object {
  "fees": Object {
    "data": Object {
      "transfer": "0.00001",
    },
    "error": null,
    "status": "SUCCESS",
  },
}
`;

exports[`ledgerStore action: RESET_LEDGER_FEES 1`] = `
Object {
  "fees": Object {
    "data": Object {
      "transfer": "0",
    },
    "error": null,
    "status": "IDLE",
  },
}
`;
